### PR TITLE
Fix over-compilation bug with Java 9, scala.ext.dir

### LINKED ISSUE #3142: [sbt 0.13.15] calling compile twice causes full compilation on JDK9
Originally reported by @smarter 

## steps

1. compile
2. compile

## problem

It causes full compilation.

## expectation

No compilation the second times.

## notes

https://gist.github.com/eed3si9n/8ee47e2e5dafaa5ef9f00833679a0a0b

### PR DESCRIPTION
**Pull request by retronym**
Fixes #3142

### CODE REVIEW

**Comment by retronym** [comment]
Before:
```
[success] Total time: 1 s, completed 1 Nov. 2017, 8:55:05 am
> compile
[info] Compiling 1 Scala source to /Users/jz/code/scratch2/target/scala-2.10/classes...
[success] Total time: 0 s, completed 1 Nov. 2017, 8:55:06 am
> compile
[info] Compiling 1 Scala source to /Users/jz/code/scratch2/target/scala-2.10/classes...
[success] Total time: 0 s, completed 1 Nov. 2017, 8:55:09 am
```

After:

```
> compile
[info] Updating {file:/Users/jz/code/scratch2/}scratch2...
[info] Resolving org.fusesource.jansi#jansi;1.4 ...
[info] Done updating.
[info] Compiling 1 Scala source and 1 Java source to /Users/jz/code/scratch2/target/scala-2.10/classes...
[success] Total time: 0 s, completed 1 Nov. 2017, 9:17:51 am
> compile
[success] Total time: 0 s, completed 1 Nov. 2017, 9:17:52 am
> sbtVersion
[info] 0.13.17-bf36f824f
```

**Comment by retronym** [comment]
The second commit fixes this particular issue. The first would be relevant on non-OpenJDK based JVMs (prior to Java 9), but is otherwise not relevant as the SBT launcher doesn't use `-Dscala.boot.class.path`.

**Comment by retronym** [comment]
I've added an extra commit that works around a similar overcompilation bug when using the upcoming Scala 2.10.7 or 2.11.12 as the project scala version, assuming that the next version of SBT will no longer set `-Dscala.ext.dirs`.

**Comment by retronym** [comment]
Build failure seems spurious:

```
[info] [debug] 	public: downloading https://repo1.maven.org/maven2/org/scala-lang/scala-compiler/2.10.4/scala-compiler-2.10.4.jar
[info] [warn] 	[FAILED     ] org.scala-lang#scala-compiler;2.10.4!scala-compiler.jar: Connection reset (248ms)
[info] [warn] 	[FAILED     ] org.scala-lang#scala-compiler;2.10.4!scala-compiler.jar: Connection reset (248ms)
[info] [warn] ==== public: tried
```

File: compile/interface/src/main/scala/xsbt/Dependency.scala [hunk 1]
```diff
@@ -63,7 +63,7 @@ final class Dependency(val global: CallbackGlobal) extends LocateClassFile with
               case Some((f, className, inOutDir)) =>
                 if (inOutDir && on.isJavaDefined) registerTopLevelSym(on)
                 f match {
-                  case ze: ZipArchive#Entry => for (zip <- ze.underlyingSource; zipFile <- Option(zip.file)) binaryDependency(zipFile, className)
+                  case ze: ZipArchive#Entry => for (zip <- ze.underlyingSource; zipFile <- Option(zip.file).filterNot(_.isDirectory)) binaryDependency(zipFile, className)
```
Comment (smarter) [inline compile/interface/src/main/scala/xsbt/Dependency.scala]: I think this is worth a comment in the source.

**Review by eed3si9n — APPROVED** [review]

File: compile/interface/src/main/scala/xsbt/Dependency.scala (hunk 1 above)
Comment (eed3si9n) [reply compile/interface/src/main/scala/xsbt/Dependency.scala]: Something like:

```scala
// This skips the Zip entry whose underlying source is a directory to make 2.10 compilation consistent with 2.12
// See https://github.com/sbt/sbt/issues/3142#issuecomment-340947086 
```

?

Comment (retronym) [reply compile/interface/src/main/scala/xsbt/Dependency.scala]: I've pushed a comment.

**Pull request merged**